lib/common/grpc-service.js

```diff
@@ -0,0 +1,230 @@
+  },
+
+  3: {
+    code: 400,
+    message: 'Bad Request'
+  },
+
+  4: {
+    code: 504,
+    message: 'Gateway Timeout'
+  },
+
+  5: {
+    code: 404,
+    message: 'Not Found'
+  },
+
+  6: {
+    code: 409,
+    message: 'Conflict'
+  },
+
+  7: {
+    code: 403,
+    message: 'Forbidden'
+  },
+
+  8: {
+    code: 429,
+    message: 'Too Many Requests'
+  },
+
+  9: {
+    code: 412,
+    message: 'Precondition Failed'
+  },
+
+  10: {
+    code: 409,
+    message: 'Conflict'
+  },
+
+  11: {
+    code: 400,
+    message: 'Bad Request'
+  },
+
+  12: {
+    code: 501,
+    message: 'Not Implemented'
+  },
+
+  13: {
+    code: 500,
+    message: 'Internal Server Error'
+  },
+
+  14: {
+    code: 503,
+    message: 'Service Unavailable'
+  },
+
+  15: {
+    code: 500,
+    message: 'Internal Server Error'
+  },
+
+  16: {
+    code: 401,
+    message: 'Unauthorized'
+  }
+};
+
+/**
+ * Service is a base class, meant to be inherited from by a "service," like
+ * BigQuery or Storage.
+ *
+ * This handles making authenticated requests by exposing a `makeReq_` function.
+ *
+ * @param {object} config - Configuration object.
+ * @param {string} config.baseUrl - The base URL to make API requests to.
+ * @param {string[]} config.scopes - The scopes required for the request.
+ * @param {object} options - [Configuration object](#/docs/?method=gcloud).
+ */
+function GrpcService(config, options) {
+  Service.call(this, config, options);
+
+  var service = config.service;
+  var apiVersion = config.apiVersion;
+  var rootDir = googleProtoFiles('..');
+
+  this.protoOpts = config.proto;
+  this.proto = grpc.load({
+    root: rootDir,
+    file: path.relative(rootDir, googleProtoFiles[service][apiVersion])
+  }).google[service][apiVersion];
+}
+
+nodeutil.inherits(GrpcService, Service);
+
+/**
+ * Make an authenticated request with gRPC.
+ *
+ * @param {object} protoOpts - The proto options.
+ * @param {string} protoOpts.service - The service name.
+ * @param {string} protoOpts.method - The method name.
+ * @param {object} reqOpts - The request options.
+ * @param {function=} callback - The callback function.
+ */
+GrpcService.prototype.request = function(protoOpts, reqOpts, callback) {
+  var self = this;
+  var proto = this.proto;
+
+  if (!this.grpcCredentials) {
+    // We must establish an authClient to give to grpc.
+    this.authClient.getCredentials(function(err) {
+      if (err) {
+        callback(err);
+        return;
+      }
+
+      var authClient = self.authClient.authClient;
+
+      self.grpcCredentials = grpc.credentials.combineChannelCredentials(
+        grpc.credentials.createSsl(),
+        grpc.credentials.createFromGoogleCredential(authClient)
+      );
+
+      self.request.call(self, protoOpts, reqOpts, callback);
+    });
+    return;
+  }
+
+  var service = new proto[protoOpts.service](
+    this.baseUrl,
+    this.grpcCredentials
+  );
+
+  service[protoOpts.method](snakeize(reqOpts), function(err, resp) {
```

Comment:
@tbetbetbe - this method is where help would be appreciated.

1) ProtoBuf.js allows you to treat snake-cased proto fields as camelcase, but I don't believe gRPC exports an option. Is that a possibility?
2) I'm manually walking all of the returned properties and converting anything buffer-like to its string value. This is to get around [this line](https://github.com/grpc/grpc/blob/d50a4ee2c257ddfadccb9d252e2b61d6ff544d96/src/node/src/common.js#L58)-- what we need here is basically `true` for the first argument to `toRaw`.
3) Errors -- is there a more proper way to get the error status code instead of the 1-16 value?